src/libraries/System.Private.CoreLib/src/System/IO/File.cs

```diff
@@ -512,64 +511,39 @@ public static Task WriteAllTextAsync(string path, string? contents, Encoding enc
             return WriteToFileAsync(path, FileMode.Create, contents, encoding, cancellationToken);
         }
 
-        public static Task<byte[]> ReadAllBytesAsync(string path, CancellationToken cancellationToken = default(CancellationToken))
+        public static async Task<byte[]> ReadAllBytesAsync(string path, CancellationToken cancellationToken = default(CancellationToken))
         {
-            if (cancellationToken.IsCancellationRequested)
-            {
-                return Task.FromCanceled<byte[]>(cancellationToken);
-            }
+            cancellationToken.ThrowIfCancellationRequested();
 
-            var fs = new FileStream(
-                path, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 1, // bufferSize == 1 used to avoid unnecessary buffer in FileStream
-                FileOptions.Asynchronous | FileOptions.SequentialScan);
-
-            bool returningInternalTask = false;
-            try
+            using (SafeFileHandle sfh = OpenHandle(path, FileMode.Open, FileAccess.Read, FileShare.Read, FileOptions.Asynchronous))
             {
                 long fileLength = 0L;
-                if (fs.CanSeek && (fileLength = fs.Length) > int.MaxValue)
+                if (sfh.CanSeek && (fileLength = RandomAccess.GetFileLength(sfh)) > Array.MaxLength)
                 {
-                    var e = new IOException(SR.IO_FileTooLong2GB);
-                    ExceptionDispatchInfo.SetCurrentStackTrace(e);
-                    return Task.FromException<byte[]>(e);
+                    throw new IOException(SR.IO_FileTooLong2GB);
                 }
-
-                returningInternalTask = true;
-                return fileLength > 0 ?
-                    InternalReadAllBytesAsync(fs, (int)fileLength, cancellationToken) :
-                    InternalReadAllBytesUnknownLengthAsync(fs, cancellationToken);
-            }
-            finally
-            {
-                if (!returningInternalTask)
+                if (fileLength == 0)
                 {
-                    fs.Dispose();
+                    return await InternalReadAllBytesUnknownLengthAsync(sfh, cancellationToken).ConfigureAwait(false);
```

Comment:
The implementation before did the disposal inside of the called functions so as to avoid the need for the try/catch in the outerloop, which then avoided the need for this extra level of async state machine.  Previously if fileLength == 0, ReadAllBytesAsync would allocate one state machine (for InternalReadAllBytesAsync), now it'll allocate two (one for ReadAllBytesAsync, one for InternalReadAllBytesUnknownLengthAsync).

Reply:
Thanks for a great explanation, I've restored the previous pattern